### PR TITLE
src/hmem_ze: select copy only engine

### PR DESCRIPTION
**Pull request by zachdworkin**
Current path does not specify an engine.
Default engine could be compute and copy.
Add a check to find a copy only engine to use for optimization.
If no copy only engine is found it will use ordinal 0 which is always available for the engine.
Signed-off-by: zdworkin <zachary.dworkin@intel.com>

### CODE REVIEW

File: src/hmem_ze.c [hunk 1]
```diff
@@ -583,11 +643,16 @@ int ze_hmem_init(void)
...
 
-	ret = ze_hmem_init_fds();
+	ze_ret = ze_hmem_init_fds();
 	if (ret)
 		goto err;
 
 	for (num_devices = 0; num_devices < count; num_devices++) {
+		ze_ret = ze_hmem_find_copy_only_engine(num_devices, &ordinals[num_devices]);
+		if (ze_ret || ordinals[num_devices] == -1)
+			goto err;
```
Comment (j-xiong) [inline src/hmem_ze.c:653]: Does copy only engine always exist? On failure, would it better to use a mixed engine than to fail?

**Review by j-xiong — COMMENTED** [review]
Please describe the reason behind the change in the commit message.

File: src/hmem_ze.c (hunk 1 above)
Comment (aingerson) [reply src/hmem_ze.c:653]: Good point, yes we should definitely fall back to a mixed engine.

**Comment by zachdworkin** [comment]
@aingerson updated all comments.

File: src/hmem_ze.c [hunk 2]
```diff
@@ -12,13 +12,13 @@
...
+ * 	   copyright notice, this list of conditions and the following
+ * 	   disclaimer.
  *
  *      - Redistributions in binary form must reproduce the above
- *        copyright notice, this list of conditions and the following
- *        disclaimer in the documentation and/or other materials
- *        provided with the distribution.
+ * 	   copyright notice, this list of conditions and the following
+ * 	   disclaimer in the documentation and/or other materials
+ * 	   provided with the distribution.
```
Comment (aingerson) [inline src/hmem_ze.c:21]: Change from tabs into spaces to remove these diffs.

File: src/hmem_ze.c [hunk 3]
```diff
@@ -46,10 +46,12 @@ static ze_context_handle_t context;
 static ze_device_handle_t devices[ZE_MAX_DEVICES];
 static ze_command_queue_handle_t cmd_queue[ZE_MAX_DEVICES];
 static int num_devices = 0;
+static ze_command_queue_group_properties_t *cq_grp_props;
```
Comment (aingerson) [inline src/hmem_ze.c:49]: I don't think this needs to be global

File: src/hmem_ze.c [hunk 4]
```diff
@@ -428,6 +442,12 @@ static int ze_hmem_dl_init(void)
 		goto err_dlclose;
 	}
 
+	libze_ops.zeDeviceGetCommandQueueGroupProperties = dlsym(libze_handle, "zeDeviceGetCommandQueueGroupProperties");
+	if (!libze_ops.zeDeviceGetCommandQueueGroupProperties) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find zeDeviceGetCommandQueueGroupProperties\n");
```
Comment (aingerson) [inline src/hmem_ze.c:447]: Can you move the string to a second line to make it a little more readable?

File: src/hmem_ze.c [hunk 5]
```diff
@@ -544,6 +564,41 @@ static void ze_hmem_dl_cleanup(void)
...
+	uint32_t cq_grp_count = 0;
+	int i;
+
+	ze_ret = ofi_zeDeviceGetCommandQueueGroupProperties(devices[device_num],
+							    &cq_grp_count, NULL);
+	if (ze_ret)
+		return ze_ret;
+
+	cq_grp_props = calloc(cq_grp_count,
+			      sizeof(ze_command_queue_group_properties_t));
```
Comment (aingerson) [inline src/hmem_ze.c:579]: sizeof(*cq_grp_props) for maintability

File: src/hmem_ze.c [hunk 6]
```diff
@@ -544,6 +564,41 @@ static void ze_hmem_dl_cleanup(void)
...
+	for (i = 0; i < cq_grp_count; i++) {
+		if (cq_grp_props[i].flags &
+		    ZE_COMMAND_QUEUE_GROUP_PROPERTY_FLAG_COPY &&
+		    !(cq_grp_props[i].flags &
+		      ZE_COMMAND_QUEUE_GROUP_PROPERTY_FLAG_COMPUTE)) {
+			break;
+		}
+	}
+
+err:
```
Comment (aingerson) [inline src/hmem_ze.c:596]: rename to out instead of err since this code is always run

File: src/hmem_ze.c [hunk 7]
```diff
@@ -544,6 +564,41 @@ static void ze_hmem_dl_cleanup(void)
...
+	free(cq_grp_props);
+	*ordinal = i == cq_grp_count ? 0 : i;
```
Comment (aingerson) [inline src/hmem_ze.c:598]: can you add a note in the commit message that says if no copy only was available it will default to ordinal 0?

**Review by aingerson — COMMENTED** [review]
And please update your commit message per Jianxin's comment.

**Comment by zachdworkin** [comment]
Updated @aingerson

**Review by aingerson — APPROVED** [review]

**Pull request merged**